### PR TITLE
fix(destructuring): rename parameter names containing curly brace.

### LINKED ISSUE #95: Unnamed object parameters
```
index (1) must be less than size (1)
  Node(PARAM_LIST): /javascript/angular2/github/modules/angular2/src/core/di/provider.js:69:15
    constructor(token, { useClass, useValue, useExisting, useFactory, deps, multi }) {
  Parent(FUNCTION ):/javascript/angular2/github/modules/angular2/src/core/di/provider.js:69:4
    constructor(token, { useClass, useValue, useExisting, useFactory, deps, multi }) {

	at com.google.common.base.Preconditions.checkElementIndex(Preconditions.java:1295)
	at com.google.common.base.Preconditions.checkElementIndex(Preconditions.java:1277)
	at com.google.common.collect.SingletonImmutableList.get(SingletonImmutableList.java:41)
```

`javascript/angular2/github/modules/angular2/src/core/di/provider.js` looks like

```
/**
     * @param token
     * @param { ?} {useClass, useValue, useExisting, useFactory, deps, multi}
     */
    constructor(token, { useClass, useValue, useExisting, useFactory, deps, multi }) {
```

I don't think that name for parameter 2 is legal.

This is the next blocker for compiling angular2 with closure compiler.

### PR DESCRIPTION
**Pull request by alexeagle**
Fixes #95

### CODE REVIEW

File: test_files/functions.sickle.ts [hunk 1]
```diff
@@ -17,3 +17,7 @@ function FunctionsTest2(a: number, b: number) {}
  * @param { number} b
  */
 function FunctionsTest3(a: number, b: number) {}
+/**
+ * @param {{a:number, b:number}} param0
```
Comment (martine) [inline test_files/functions.sickle.ts]: Just for more test coverage, you might throw in two destructuring params here, or a destructure of the form

function foo({a, b}, [c, d])

**Comment by martine** [comment]
LGTM either way.

**Comment by alexeagle** [comment]
Nice, I did not realize arrays could be destructured as well.

Now I encounter an error, this is the first array type annotation and needs
to be converted to Array<number> from number[].

````
/Users/alexeagle/Projects/sickle/test_files/functions.tr.js:21: ERROR - Bad
type annotation. expected closing }
 * @param { number[]} param1
                  ^

/Users/alexeagle/Projects/sickle/test_files/functions.tr.js:21: ERROR - Bad
type annotation. expecting a variable name in a @param tag
 * @param { number[]} param1
                  ^

2 error(s), 0 warning(s)
```

On Mon, Mar 14, 2016 at 12:18 PM Evan Martin <notifications@github.com>
wrote:

> LGTM either way.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/angular/sickle/pull/96#issuecomment-196483165>.
>

**Comment by martine** [comment]
Ah, good point.  Skip the array for now, it's a separate issue.

**Pull request merged**